gibbon-compiler/src/Gibbon/Passes/CalculateBounds.hs

```diff
@@ -0,0 +1,158 @@
+              (ls', regEnvs) <- unzip <$> mapM go ls
+              return (MkProdE ls', M.unionsWith max regEnvs)
+            LetE (v, locs, ty0, bind) bod -> do
+              (bind', regEnv') <- go bind
+              let venv' = M.insert v ty0 (vEnv env2)
+              (bod', regEnv'') <- calculateBoundsExp ddefs env2 { vEnv = venv' } szEnv locEnv regEnv bod
+              return (LetE (v, locs, ty0, bind') bod', M.unionWith max regEnv' regEnv'')
+            CaseE ex2 cases -> do
+              (cases', res) <-
+                unzip
+                  <$> mapM
+                        (\(dcon :: DataCon, vlocs :: [(Var, LocVar)], bod :: Exp2) -> do
+                          -- let offsets = M.fromList 
+                          --       . tail -- remove tag
+                          --       . scanl1 (\(_, s1) (v2, s2) -> (v2, s1+s2)) -- accumulate offsets
+                          --       . ((undefined, BoundedSize 1):)  -- add size for tag
+                          --       . zip (map snd vlocs) -- take locations
+                          --       . map (maybe Undefined BoundedSize . sizeOfTy) -- map to our region size type
+                          --       $ lookupDataCon ddefs dcon -- find ddef)
+                          -- * NOTE: After thinking, no location can be allocated from offset to an internal argument
+                          -- of a data constructor, except jumps so we can skip this analysis.
+                          (bod', re) <- go bod
+                          return ((dcon, vlocs, bod'), re)
+                        )
+                        cases
+              return (CaseE ex2 cases', M.unionsWith max res)
+            Ext ext -> case ext of
+              LetRegionE reg bod -> do
+                (bod', re) <- go bod
+                traceM $ ">> RegionSize: " ++ show reg ++ " -> " ++ show (re M.! regionToVar reg)
+                return (Ext $ LetRegionE (AnalyzedRegion reg (re M.! regionToVar reg)) bod', re)
+              LetParRegionE{}        -> todo
+              LetLocE loc locExp ex1 -> do
+                -- * NOTE: jumps are only necessary for route ends, skipping them.
+                if "jump_" `isPrefixOf` fromVar loc
+                  then do
+                    (ex1', re') <- go ex1
+                    return (Ext $ LetLocE loc locExp ex1', re')
+                  else do
+                    let (re, rs) = getRegionSize locExp
+                    let regEnv'  = M.insertWith (+) re rs regEnv
+                    let le'      = M.insert loc re locEnv
+                    (ex1', re') <- calculateBoundsExp ddefs env2 szEnv le' regEnv' ex1
+                    return (Ext $ LetLocE loc locExp ex1', re')
+              RetE _locs v -> do
+                (_, re) <- go (VarE v)
+                return (ex, re)
+              FromEndE{}         -> todo
+              AddFixed{}         -> todo
+              GetCilkWorkerNum{} -> todo
+              LetAvail{}         -> todo
```

Comment:
Need to recurse here.

Reply:
Can you help me with more information about this?

Reply:
Sorry. I wanted to say that `LetAvail` has a body expression and the pass should recurse on it as well.